File: Python/Codes/Practice/LeetCodeCN/771 Jewels And Stones.py


class Solution:
	def numJewelsInStones(self, J, S):
		"""
		:type J: str
		:type S: str
		:rtype: int
		"""
		count = 0
		cachedTbl = [False] * 256
		for i in J:
			cachedTbl[ord(i)] = True

		for i in S:
			if cachedTbl[ord(i)]:
				count+=1

		return count



if __name__ == '__main__':

	s = Solution()    
	print(s.numJewelsInStones("aA","aAAbbbb"));
	print(s.numJewelsInStones("z","ZZ"));
	print(s.numJewelsInStones("",""));
	print(s.numJewelsInStones("Abcdf","afca1wdfAffaFDab"));

File: Cpp/771 Jewels And Stones.cpp
#include <gtest\gtest.h>
#include <unordered_set>
using namespace std;


int numJewelsInStones(string J, string S) 
{
	unordered_set<char> tblJewel(J.begin(),J.end());
	int count = 0;

	for (auto i : S)
	{
		if (tblJewel.find(i) != tblJewel.end())
		{
			count++;
		}
	}

	return count;
}

int numJewelsInStones2(string J, string S)
{
	vector<bool> tblJewel(256, false);
	int count = 0;
	for (auto i : J)
	{
		tblJewel[i] = true;
	}

	for (auto i : S)
	{
		if (tblJewel[i])
		{
			count++;
		}
	}

	return count;
}

TEST(LeetCodeCN, tNumJewelsInStones)
{
	ASSERT_EQ(numJewelsInStones2("aA", "aAAbbbb"), 3);
	ASSERT_EQ(numJewelsInStones2("z", "ZZ"), 0);
	ASSERT_EQ(numJewelsInStones2("", ""), 0);
	ASSERT_EQ(numJewelsInStones2("Abcdf", "afca1wdfAffaFDab"), 8);
}